### PR TITLE
Add a utility function to filter out elements that are not of a certain type

### PR DESCRIPTION
**Pull request by naoyam**
This adds a utility function that I think can reduce repetition. It's not yet completed, but I'd like to have feedback before going further.

We often have code like this:

```
std::vector<Val*> values;
for (auto x: values) {
  // do something only for TensorView vals
  if (x->getValType() == ValType::TensorView) {
   TensorView* x_tv = x->as<TensorView>();
   ...
  }
}
```

This PR adds `ir_utils::filterVals`, which can be used to simplify code patterns like the above:

```
for (auto x: filterVals<TensorView>(values)) {
  // x is TensorView*
 ... 
}
```

For this, I added `static constexpr ValType type` to the `Val` classes (not yet completed; only added to `IterDomain` and `TensorView` for now).

Let me know if you have any feedback. Thanks!

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/ir_utils.h [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+namespace jit {
+namespace fuser {
+
+class Statement;
+
+namespace ir_utils {
+
+template <
+    typename FilterType,
+    template <typename, typename...> typename Container,
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/ir_utils.h]: idea: instead of trying to fit the container type, how about we pass it more generically and use std::begin(c) ... std::end(c) to iterate it? (that's what range based for does anyway) this way it would work even with plain C arrays.

https://en.cppreference.com/w/cpp/iterator/begin

File: torch/csrc/jit/codegen/cuda/ir_utils.h [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    typename ElementType>
+Container<FilterType*> filterVals(const Container<ElementType*>& container) {
+  Container<FilterType*> out;
+  for (auto& s : container) {
+    if (s->getValType() == FilterType::type) {
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/ir_utils.h]: since you're interested in a better notation, would you be interested in adding another building block to the IR root class? ie. add this to Statement:

```cpp
 template<class T>
 bool IsA() const {
 return dynamic_cast<const T*>(this) != nullptr;
 }
```

this is a counterpart to Statement::as<T> that I have on my todo list, and would allow us to say:

`if (s->isA<FilterType>()) { ... }`

... and also eventually eliminate the need for ValType altogether.

File: torch/csrc/jit/codegen/cuda/ir_utils.h [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+namespace ir_utils {
+
+template <
+    typename FilterType,
+    template <typename, typename...> typename Container,
+    typename ElementType>
+Container<FilterType*> filterVals(const Container<ElementType*>& container) {
+  Container<FilterType*> out;
+  for (auto& s : container) {
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/ir_utils.h]: nit: using the explicit root type here woudl avoid the need for the `template as<FilterType>` below

File: torch/csrc/jit/codegen/cuda/ir_utils.h [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+class Statement;
+
+namespace ir_utils {
+
+template <
+    typename FilterType,
+    template <typename, typename...> typename Container,
+    typename ElementType>
+Container<FilterType*> filterVals(const Container<ElementType*>& container) {
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/ir_utils.h]: why does it have to return the same container as the source?

File: torch/csrc/jit/codegen/cuda/ir_utils.h (hunk 2 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_utils.h]: That looks nice, but one thing I didn't want to change is that the type checking is currently always done with `getValType()`. I think we can replace that with something like `IsA`, but I wonder whether `ValType` values and classes have one-to-one relationship. For example, would there be a case where there could be two different classes that are a subclass of `Val` AND have the same `ValType` value. I don't think that's the case now, but is this something that could happen in the future?

File: torch/csrc/jit/codegen/cuda/ir_utils.h (hunk 4 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_utils.h]: Just for simplicity.

File: torch/csrc/jit/codegen/cuda/ir_utils.h (hunk 3 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_utils.h]: Sorry, can you give me more details?

File: torch/csrc/jit/codegen/cuda/ir_utils.h (hunk 1 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_utils.h]: I'd love to know what that would look like. Can you show some sketch?

File: torch/csrc/jit/codegen/cuda/ir_utils.h [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+      ++input_it_;
+    }
+  }
+};
+
+template <typename FilterType, typename InputIt>
+class FilterValContainer {
+ public:
+  using value_type = FilterType*;
+  using const_iterator = FilterIterator<FilterType, InputIt>;
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/ir_utils.h:77]: nitpick: Wouldn't we also want a non-const iterator?

File: torch/csrc/jit/codegen/cuda/ir_utils.h (hunk 2 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/ir_utils.h]: I think switch to a more consistent `IsA` makes sense to me (uniformed API).
Can't we have specialization of `isA` so that we cannot necessarily do dynamic cast? that seems to also resolve the restriction on the one-to-one mapping.

```
 template<class T>
 bool IsA() const {
 return dynamic_cast<const T*>(this) != nullptr;
 }

 template<ValType T>
 bool IsA() const {
 return getValType() == T;
 }
```

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/ir_utils.h]: Using `IsA<>()` with the non-mapping filter and return would still require us to add an extra template parameter in the filterVals, (one for filter criterion used in the `IsA`, one to cast the return type `as`). Might bother people :shrug:

**Review by jjsjann123 — COMMENTED** [review]
I like the updated version a lot :+1:

File: torch/csrc/jit/codegen/cuda/ir_utils.h (hunk 2 above)
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/ir_utils.h]: @naoyam: a potential mismatch between the tag (`ValType`) and the actual type is one of the reasons I'm wary of using tags  , especially when they duplicate the type hierarchy 1:1 (like we do today). Since we're using a hierarchy to model the taxonomy of nodes, I think it's both easier and safer to use the types themselves. Another benefit of using the types (through a notation like `isA<T>`), is that you can query for abstract (non-leaf) types too, ex. `isA<Expr>` or `isA<Val>`

@jjsjann123 : as I explained above, the main benefit of `isA<T>` is not the notation itself, but the fact that it avoids tags altogether. So a version which tries to mix tags and types seems confusing at best and I'd recommend against it.

I'm not suggesting we should go as far as eliminating tags in this change. It seems an opportunity for adding `isA<T>` since it fits well with the proposed utility, but it's not really critical to this PR.

File: torch/csrc/jit/codegen/cuda/ir_utils.h (hunk 3 above)
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/ir_utils.h]: I just meant `for (Statement* s : container) ...`.

File: torch/csrc/jit/codegen/cuda/ir_utils.h (hunk 4 above)
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/ir_utils.h]: Wouldn't be simpler (for the non-iterator version) to just return a `std::vector<FilterType*>`? That avoids the whole complication with trying to deduce the container type, and leads to more predictable behavior too (we wouldn't necessarily want to use a std::list or std::deque to return the results just to match the original container)

File: torch/csrc/jit/codegen/cuda/ir_utils.h [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+      if ((*input_it_)->getValType() == FilterType::type) {
+        break;
+      }
+      ++input_it_;
+    }
+  }
+};
+
+template <typename FilterType, typename InputIt>
+class FilterValContainer {
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/ir_utils.h]: We should probably rename this to say "view" to better reflect what it is (it's not really a container since it doesn't own the values). `FilteredView` ?

Also, `InputIt` -> just `Iterator`? (technically it's not an input iterator but rather a forward iterator)

File: torch/csrc/jit/codegen/cuda/ir_utils.h (hunk 5 above)
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/ir_utils.h:77]: Do we want to modify the original container through the filtered view?

File: torch/csrc/jit/codegen/cuda/ir_utils.h [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+  const_iterator begin() const {
+    return const_iterator(input_it_, last_);
+  }
+
+  const_iterator end() const {
+    return const_iterator(last_, last_);
+  }
+
+ private:
+  InputIt input_it_;
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/ir_utils.h]: 1. `const` for the data members?

2. Use more precise names for the iterators: `begin_` and `end_` ? (in particular, last != end so we should avoid using that name)

File: torch/csrc/jit/codegen/cuda/ir_utils.h [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+};
+
+template <typename FilterType, typename InputIt>
+auto filterVals(InputIt first, InputIt last) {
+  return FilterValContainer<FilterType, InputIt>(first, last);
+}
+
+template <typename FilterType, typename ContainerType>
+auto filterVals(const ContainerType& inputs) {
+  return filterVals<FilterType>(inputs.begin(), inputs.end());
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/ir_utils.h]: nit: `cbegin`, `cend`

File: torch/csrc/jit/codegen/cuda/ir_utils.h [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+  Container<FilterType*> out;
+  for (auto& s : container) {
+    if (s->getValType() == FilterType::type) {
+      out.push_back(s->template as<FilterType>());
+    }
+  }
+  return out;
+}
+#else
+template <typename FilterType, typename InputIt>
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/ir_utils.h]: `InputIt` -> `IteratorType` or just `Iterator` ?

File: torch/csrc/jit/codegen/cuda/ir_utils.h [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+  bool operator==(const FilterIterator& other) const {
+    return input_it_ == other.input_it_ && last_ == other.last_;
+  }
+
+  bool operator!=(const FilterIterator& other) const {
+    return !(*this == other);
+  }
+
+ private:
+  InputIt input_it_;
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/ir_utils.h]: please separate the data members from private functions into separate `private` sections

File: torch/csrc/jit/codegen/cuda/ir_utils.h [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+ public:
+  FilterIterator(InputIt first, InputIt last) : input_it_(first), last_(last) {
+    advance();
+  }
+
+  FilterType* operator*() const {
+    return (*input_it_)->template as<FilterType>();
+  }
+
+  FilterIterator& operator++() {
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/ir_utils.h:34]: we need a post increment implementation too

File: torch/csrc/jit/codegen/cuda/ir_utils.h [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+    return input_it_ == other.input_it_ && last_ == other.last_;
+  }
+
+  bool operator!=(const FilterIterator& other) const {
+    return !(*this == other);
+  }
+
+ private:
+  InputIt input_it_;
+  InputIt last_;
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/ir_utils.h]: 1. rename input_it_ -> current_ ?
2. as I mentioned in another comment, we should rename last_ -> end_

File: torch/csrc/jit/codegen/cuda/ir_utils.h [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+  }
+
+  FilterIterator& operator++() {
+    ++input_it_;
+    advance();
+    return *this;
+  }
+
+  bool operator==(const FilterIterator& other) const {
+    return input_it_ == other.input_it_ && last_ == other.last_;
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/ir_utils.h]: I'd actually just `TORCH_INTERNAL_ASSERT(last_ == other.last_)` - if someone is trying to compare fitered iterators coming from different filtered views it's most likely an error and we should raise it as such

File: torch/csrc/jit/codegen/cuda/ir_utils.h [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+    advance();
+  }
+
+  FilterType* operator*() const {
+    return (*input_it_)->template as<FilterType>();
+  }
+
+  FilterIterator& operator++() {
+    ++input_it_;
+    advance();
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/ir_utils.h:36]: could we use `std::find_if` instead?

File: torch/csrc/jit/codegen/cuda/ir_utils.h (hunk 7 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_utils.h]: Thanks. Done.

File: torch/csrc/jit/codegen/cuda/ir_utils.h (hunk 8 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_utils.h]: Changed.

File: torch/csrc/jit/codegen/cuda/ir_utils.h (hunk 9 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_utils.h]: Renamed to `Iterator`.

File: torch/csrc/jit/codegen/cuda/ir_utils.h (hunk 10 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_utils.h]: Done.

File: torch/csrc/jit/codegen/cuda/ir_utils.h (hunk 12 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_utils.h]: Done.

File: torch/csrc/jit/codegen/cuda/ir_utils.h (hunk 13 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_utils.h]: That's a good idea. Done.

File: torch/csrc/jit/codegen/cuda/ir_utils.h (hunk 6 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_utils.h]: Thanks for the suggestion. Renamed to `FilteredView`.

File: torch/csrc/jit/codegen/cuda/ir_utils.h (hunk 5 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_utils.h:77]: There may be some use cases, but I suspect we don't need that at this time. I left a related comment for a future extension.

File: torch/csrc/jit/codegen/cuda/ir_utils.h (hunk 2 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_utils.h]: I agree that we could get rid of the tag and just use the runtime type information would be more robust. However, I'd prefer to keep using the tag for type checking at least in this PR as changing that is not the purpose. I'll create a separate issue to consider.

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_utils.h]: See #268

File: torch/csrc/jit/codegen/cuda/ir_utils.h (hunk 14 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_utils.h:36]: Thanks for the suggestion. Simplified the implementation of `advance` with `find_if`.

File: torch/csrc/jit/codegen/cuda/ir_utils.h (hunk 11 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_utils.h:34]: Done

**Comment by naoyam** [comment]
I actually changed the type checking to use `dynamic_cast` instead of `getValType()` as it eventually does `dynamic_cast`. I think it makes more sense.

**Review by jjsjann123 — APPROVED** [review]
LGTM

File: torch/csrc/jit/codegen/cuda/ir_utils.h [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+ public:
+  FilterIterator(Iterator begin, Iterator end) : current_(begin), end_(end) {
+    advance();
+  }
+
+  FilterType* operator*() const {
+    return (*current_)->template as<FilterType>();
+  }
+
+  FilterType* operator->() const {
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/ir_utils.h:30]: nice, good catch!

File: torch/csrc/jit/codegen/cuda/ir_utils.h [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+  }
+
+  FilterIterator& operator++() {
+    ++current_;
+    advance();
+    return *this;
+  }
+
+  FilterIterator operator++(int) {
+    auto before_increment = *this;
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/ir_utils.h]: nit: `const` ?

File: torch/csrc/jit/codegen/cuda/ir_utils.h [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+        end_ == other.end_,
+        "Comparing two FilteredViews that originate from different containers");
+    return current_ == other.current_;
+  }
+
+  bool operator!=(const FilterIterator& other) const {
+    return !(*this == other);
+  }
+
+ private:
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/ir_utils.h:58]: please place the private state section last (public methods, private methods, private state)

**Review by tlemo — APPROVED** [review]

File: torch/csrc/jit/codegen/cuda/ir_utils.h (hunk 17 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_utils.h:58]: Done

File: torch/csrc/jit/codegen/cuda/ir_utils.h (hunk 16 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_utils.h]: Done

File: test/cpp/jit/test_gpu.cpp [hunk 18]
```diff
@@ -781,6 +782,50 @@ void testGPU_FusionTensor() {
...
+                 from ../torch/csrc/jit/ir/ir.h:3,
+                 from ../test/cpp/jit/test_base.h:5,
+                 from ../test/cpp/jit/test_gpu.cpp:3:
+/usr/include/c++/8/bits/stl_vector.h:543:2: note: candidate: ‘template<class _InputIterator, class> std::vector<_Tp, _Alloc>::vector(_InputIterator, _InputIterator, const allocator_type&)’
+  vector(_InputIterator __first, _InputIterator __last,
+  */
+#if 0
+  std::vector<TensorView*> tvs(
+      ir_utils::filterVals<TensorView>(vals).begin(),
+      ir_utils::filterVals<TensorView>(vals).end());
```
Comment (naoyam) [inline test/cpp/jit/test_gpu.cpp]: @tlemo Do you know why this doesn't work?

Comment (jjsjann123) [reply test/cpp/jit/test_gpu.cpp]: Look at constructor (5) in https://en.cppreference.com/w/cpp/container/vector/vector
Your const_iterator needs to satisfy `LegacyInputIterator`... So you would need `difference_type` and bunch of that...

File: test/cpp/jit/test_gpu.cpp [hunk 19]
```diff
@@ -781,6 +782,40 @@ void testGPU_FusionTensor() {
...
+  TORCH_CHECK(floats.size() == 1);
+  TORCH_CHECK(floats[0] == scalar0);
+
+  std::vector<Int*> ints(
+      ir_utils::filterVals<Int>(vals).begin(),
+      ir_utils::filterVals<Int>(vals).end());
+  TORCH_CHECK(ints.size() == 2);
+  TORCH_CHECK(ints[0] == scalar1);
+  TORCH_CHECK(ints[1] == scalar2);
+  return;
```
Comment (tlemo) [inline test/cpp/jit/test_gpu.cpp:816]: nit: `return` not needed

File: test/cpp/jit/test_gpu.cpp [hunk 20]
```diff
@@ -781,6 +782,40 @@ void testGPU_FusionTensor() {
...
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  auto tv0 = makeDummyTensor(1);
+  auto tv1 = makeDummyTensor(1);
+  auto scalar0 = new Float(0);
+  auto scalar1 = new Int(0);
+  auto scalar2 = new Int(1);
+
+  std::vector<Val*> vals = {tv0, scalar0, tv1, scalar1, scalar2};
```
Comment (tlemo) [inline test/cpp/jit/test_gpu.cpp:795]: can you please make vals `const`?

File: test/cpp/jit/test_gpu.cpp [hunk 21]
```diff
@@ -781,6 +782,40 @@ void testGPU_FusionTensor() {
...
+  TORCH_CHECK(tvs[0] == tv0);
+  TORCH_CHECK(tvs[1] == tv1);
+
+  std::vector<Float*> floats(
+      ir_utils::filterVals<Float>(vals).begin(),
+      ir_utils::filterVals<Float>(vals).end());
+  TORCH_CHECK(floats.size() == 1);
+  TORCH_CHECK(floats[0] == scalar0);
+
+  std::vector<Int*> ints(
```
Comment (tlemo) [inline test/cpp/jit/test_gpu.cpp:810]: one more test case please: a filtering which returns an empty list, ex:

```
for (auto ptr : ir_utils::filterVals<Expr>(vals)) {
 TORCH_CHECK(false, "Not expecting any results");
}

```

**Review by tlemo — APPROVED** [review]

**Comment by naoyam** [comment]
I renamed the filter function to `filterByType` as it is actually no longer specific to `Val`. It just checks the runtime type of each element and filters out anything that is of that type.

The example I gave first still looks mostly the same. We replace this:

```
std::vector<Val*> values;
for (Val* x: values) {
  // do something only for TensorView vals
  if (x->getValType() == ValType::TensorView) {
   TensorView* x_tv = x->as<TensorView>();
   ...
  }
}
```

with:

```
for (TensorView* x: filterByType<TensorView>(values)) {
  // x is TensorView*
 ... 
}
```

**Comment by naoyam** [comment]
@tlemo Sorry, I missed your last comments. I'll update the test with a separate PR.
